ww4ff/run/train.py

```diff
@@ -10,93 +9,105 @@
-                except StopIteration:
-                    break
-                example = trim([example])[0]
-                pred = engine.infer(example.audio_data.to(device))
-                label = example.contains_wake_word
-                conf_matrix.increment(pred, label)
-                pbar.set_postfix(dict(measure=f'{conf_matrix.mcc}'))
+        pbar = tqdm(dataset, desc=prefix)
+        for idx, batch in enumerate(pbar):
+            batch = batch.to(device)  # type: ClassificationBatch
+            pred = engine.infer(batch.audio_data.to(device).squeeze(0))
+            label = batch.labels.item()
+            conf_matrix.increment(pred < num_labels - 1, label < num_labels - 1)
+            if idx % 10 == 9:
+                pbar.set_postfix(dict(mcc=f'{conf_matrix.mcc}', c=f'{conf_matrix}'))
         logging.info(f'{conf_matrix}')
-        writer.add_scalar(f'{prefix}/Metric/mcc', conf_matrix.mcc, epoch_idx)
-        if dataset.set_type == DatasetType.DEV:
+        if save and not args.eval:
+            writer.add_scalar(f'{prefix}/Metric/mcc', conf_matrix.fp, epoch_idx)
             ws.increment_model(model, conf_matrix.mcc)
 
     apb = ArgumentParserBuilder()
     apb.add_options(opt('--model', type=str, choices=model_names(), default='las'),
                     opt('--workspace', type=str, default=str(Path('workspaces') / 'default')),
+                    opt('--load-weights', action='store_true'),
+                    opt('--vocab', type=str, nargs='+', default=[' hey', 'fire', 'fox']),
                     opt('--eval', action='store_true'))
     args = apb.parser.parse_args()
 
-    ws = Workspace(Path(args.workspace))
+    num_labels = len(args.vocab) + 1
+    ws = Workspace(Path(args.workspace), delete_existing=not args.eval)
+    writer = ws.summary_writer
     set_seed(SETTINGS.training.seed)
     ww = SETTINGS.training.wake_word
     logging.info(f'Using {ww}')
-    loader = FlatWavDatasetLoader()
-    ds_kwargs = dict(sr=SETTINGS.audio.sample_rate, mono=SETTINGS.audio.use_mono)
+    loader = WakeWordDatasetLoader()
+    ds_kwargs = dict(sr=SETTINGS.audio.sample_rate, mono=SETTINGS.audio.use_mono, words=args.vocab)
     ww_train_ds, ww_dev_ds, ww_test_ds = loader.load_splits(SETTINGS.dataset.dataset_path, **ds_kwargs)
     print_stats('Wake word dataset', ww_train_ds, ww_dev_ds, ww_test_ds)
 
     sr = SETTINGS.audio.sample_rate
     wind_sz = int(SETTINGS.training.eval_window_size_seconds * sr)
     stri_sz = int(SETTINGS.training.eval_stride_size_seconds * sr)
-    ww_train_ds = WakeWordTrainingDataset(ww_train_ds, ww)
-    ww_dev_ds = WakeWordEvaluationDataset(WakeWordTrainingDataset(ww_dev_ds, ww), wind_sz, stri_sz)
-    ww_test_ds = WakeWordEvaluationDataset(WakeWordTrainingDataset(ww_test_ds, ww), wind_sz, stri_sz)
+    ww_dev_pos_ds = ww_dev_ds.filter(lambda x: x.compute_frame_labels(args.vocab), clone=True)
+    ww_dev_neg_ds = ww_dev_ds.filter(lambda x: not x.compute_frame_labels(args.vocab), clone=True)
```

Comment:
if I understood correctly, "hey xxxx" will still fall into pos case as compute_frame_labels non-empty dict.
Can you double check?

Reply:
Yup, that's correct. For the negative set, we want speech where there are no positive labels at all.